File: behavioral/dart/mediator_structured/interfaces.dart
abstract class Mediator {
  void sendMessage(String message, String sender);
}

abstract class Colleague {
  final Mediator _mediator;
  final String _name;
  
  Colleague(this._mediator, this._name);
  
  void send(String message) {
    _mediator.sendMessage(message, _name);
  }
  
  void receive(String message);
  
  String get name => _name;
}

File: behavioral/typescript/mediator_structured/interfaces.ts
export interface Mediator {
  sendMessage(message: string, sender: string): void;
}

export abstract class Colleague {
  protected mediator: Mediator;
  protected name: string;
  
  constructor(mediator: Mediator, name: string) {
    this.mediator = mediator;
    this.name = name;
  }
  
  send(message: string): void {
    this.mediator.sendMessage(message, this.name);
  }
  
  abstract receive(message: string): void;
  
  getName(): string {
    return this.name;
  }
}
